vivisect/parsers/elf.py

```diff
@@ -231,119 +351,362 @@ def loadElfIntoWorkspace(vw, elf, filename=None, baseaddr=None):
+        vw.setVaSetRow('PointersFromFile', (va, tva, fname, pname))
+
     return fname
+
+def applyRelocs(elf, vw, addbase=False, baseaddr=0):
+    # process relocations / strings (relocs use Dynamic Symbols)
+    arch = arch_names.get(elf.e_machine)
+    relocs = elf.getRelocs()
+    logger.debug("reloc len: %d", len(relocs))
+    for r in relocs:
+        rtype = Elf.getRelocType(r.r_info)
+        rlva = r.r_offset
+        if addbase:
+            rlva += baseaddr
+        try:
+            # If it has a name, it's an externally
+            # resolved "import" entry, otherwise, just a regular reloc
+            name = r.getName()
+            dmglname = demangle(name)
+            logger.debug('relocs: 0x%x: %r  (%r)', rlva, dmglname, name)
+            if arch in ('i386', 'amd64'):
+                if name:
+                    if rtype == Elf.R_X86_64_IRELATIVE:
+                        # before making import, let's fix up the pointer as a BASEPTR Relocation
+                        ptr = r.r_addend
+                        vw.addRelocation(rlva, vivisect.RTYPE_BASEPTR, ptr)
+                        logger.info('Reloc: R_X86_64_IRELATIVE 0x%x', rlva)
+
+                    if rtype in (Elf.R_386_JMP_SLOT, Elf.R_X86_64_GLOB_DAT, Elf.R_X86_64_IRELATIVE):
+                        logger.info('Reloc: making Import 0x%x (name: %s/%s) ', rlva, name, dmglname)
+                        vw.makeImport(rlva, "*", dmglname)
+                        vw.setComment(rlva, name)
+
+                    elif rtype in (Elf.R_386_32, Elf.R_386_COPY):
+                        pass
+
+                    else:
+                        logger.warn('unknown reloc type: %d %s (at %s)', rtype, name, hex(rlva))
+                        logger.info(r.tree())
+                       
+                else:
+                    symidx = Elf.getRelocSymTabIndex(r.r_info)
+                    sym = elf.getDynSymbol(symidx)
+                    ptr = sym.st_value
+
+                    if rtype in (Elf.R_386_RELATIVE, ): # R_X86_64_RELATIVE is the same number
+                        ptr = vw.readMemoryPtr(rlva)
+                        logger.info('R_386_RELATIVE: adding Relocation 0x%x -> 0x%x (name: %s) ', rlva, ptr, dmglname)
+                        vw.addRelocation(rlva, vivisect.RTYPE_BASEPTR, ptr)
+
+                    elif rtype == Elf.R_X86_64_IRELATIVE:
+                        # first make it a relocation that is based on the imagebase
+                        ptr = r.r_addend
+                        logger.info('R_X86_64_IRELATIVE: adding Relocation 0x%x -> 0x%x (name: %r %r) ', rlva, ptr, name, dmglname)
+                        vw.addRelocation(rlva, vivisect.RTYPE_BASEPTR, ptr)
+
+                        # next get the target and find a name, since the reloc itself doesn't have one
+                        # FIXME: this doesn't actually do anything, since naming of PLT functions doesn't happen until later
```

Comment:
Will this ever do anything, and if the answer is no, why does it exist (since PLT naming happens later)?

Reply:
it actually does.  removed the comment.